File: rust/src/cert.rs
use stedy::{
    ed25519_sign, ed25519_verify, Ed25519KeyPair, Ed25519PublicKey, Ed25519Signature, Vec,
    ED25519_PUBLIC_KEY_SIZE,
};

use crate::Error;

fn calculate_subject(public_key: &Ed25519PublicKey, data: Option<&[u8]>) -> Vec<u8> {
    let mut subject = data.unwrap_or_default().to_vec();
    let max_size = usize::MAX - ED25519_PUBLIC_KEY_SIZE;
    if subject.len() > max_size {
        subject.truncate(max_size);
    }
    subject.extend_from_slice(public_key);
    subject
}

pub fn certify(
    our_identity_key_pair: &Ed25519KeyPair,
    their_identity_key: &Ed25519PublicKey,
    data: Option<&[u8]>,
) -> Result<Ed25519Signature, Error> {
    let subject = calculate_subject(their_identity_key, data);
    ed25519_sign(our_identity_key_pair, &subject).or(Err(Error::Certification))
}

pub fn verify(
    owner_identity_key: &Ed25519PublicKey,
    certifier_identity_key: &Ed25519PublicKey,
    signature: &Ed25519Signature,
    data: Option<&[u8]>,
) -> Result<(), Error> {
    let subject = calculate_subject(owner_identity_key, data);
    ed25519_verify(certifier_identity_key, signature, &subject).or(Err(Error::Verification))
}

File: typescript/src/cert.ts
import {
  autograph_certify,
  autograph_verify,
  autograph_signature_size
} from './clib'

export const createSignature = () => new Uint8Array(autograph_signature_size())

export const certify = (
  ourIdentityKeyPair: Uint8Array,
  theirIdentityKey: Uint8Array,
  data?: Uint8Array
) => {
  const signature = createSignature()
  const success = data
    ? autograph_certify(
        signature,
        ourIdentityKeyPair,
        theirIdentityKey,
        data,
        data.length
      )
    : autograph_certify(
        signature,
        ourIdentityKeyPair,
        theirIdentityKey,
        new Uint8Array(),
        0
      )
  if (!success) {
    throw new Error('Certification failed')
  }
  return signature
}

export const verify = (
  ownerIdentityKey: Uint8Array,
  certifierIdentityKey: Uint8Array,
  signature: Uint8Array,
  data?: Uint8Array
) => {
  const verified = data
    ? autograph_verify(
        ownerIdentityKey,
        certifierIdentityKey,
        signature,
        data,
        data.length
      )
    : autograph_verify(
        ownerIdentityKey,
        certifierIdentityKey,
        signature,
        new Uint8Array(),
        0
      )
  return verified
}
